contracts/governance/utils/VotesConfidential.sol

```diff
@@ -0,0 +1,209 @@
+     * Votes that have not been delegated are still part of total supply, even though they would not participate in a
+     * vote.
+     *
+     * Requirements:
+     *
+     * - `timepoint` must be in the past. If operating using block numbers, the block must be already mined.
+     */
+    function getPastTotalSupply(uint256 timepoint) public view virtual returns (euint64) {
+        return _totalCheckpoints.upperLookupRecent(_validateTimepoint(timepoint));
+    }
+
+    /**
+     * @dev Returns the current total supply of votes as an encrypted uint64 (euint64). Must be implemented
+     * by the derived contract.
+     */
+    function totalSupply() public view virtual returns (euint64);
+
+    /// @dev Returns the current total supply of votes.
+    function getCurrentTotalSupply() public view virtual returns (euint64) {
+        return _totalCheckpoints.latest();
+    }
+
+    /// @dev Returns the delegate that `account` has chosen.
+    function delegates(address account) public view virtual returns (address) {
+        return _delegatee[account];
+    }
+
+    /// @dev Delegates votes from the sender to `delegatee`.
+    function delegate(address delegatee) public virtual {
+        _delegate(msg.sender, delegatee);
+    }
+
+    /// @dev Delegates votes from `delegator` to `delegatee`.
+    function delegateBySig(
+        address delegator,
+        address delegatee,
+        uint256 nonce,
+        uint256 expiry,
+        bytes memory signature
+    ) public virtual {
+        if (block.timestamp > expiry) {
+            revert VotesExpiredSignature(expiry);
+        }
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                delegator,
+                _hashTypedDataV4(keccak256(abi.encode(DELEGATION_TYPEHASH, delegatee, nonce, expiry))),
+                signature
+            )
+        ) {
+            revert VotesInvalidSignature(delegator);
+        }
+        _useCheckedNonce(delegator, nonce);
+        _delegate(delegator, delegatee);
+    }
+
+    /**
+     * @dev Delegate all of `account`'s voting units to `delegatee`.
+     *
+     * Emits events {IVotes-DelegateChanged} and {IVotes-DelegateVotesChanged}.
+     */
+    function _delegate(address account, address delegatee) internal virtual {
+        address oldDelegate = delegates(account);
+        _delegatee[account] = delegatee;
+
+        emit DelegateChanged(account, oldDelegate, delegatee);
+        _moveDelegateVotes(oldDelegate, delegatee, _getVotingUnits(account));
+    }
+
+    /**
+     * @dev Transfers, mints, or burns voting units. To register a mint, `from` should be zero. To register a burn, `to`
+     * should be zero. Total supply of voting units will be adjusted with mints and burns.
+     */
+    function _transferVotingUnits(address from, address to, euint64 amount) internal virtual {
+        if (from == address(0) || to == address(0)) {
+            _push(_totalCheckpoints, totalSupply());
```

Comment:
This is a VERY different behavior from the non-confidential Votes.sol.

Is that on purpose? Is there a security issue in replicating Votes's behavior ?

Reply:
It's quite expensive to perform the total supply math twice (external calls), as is done in normal votes. I prefer to inherit the total supply from the token contract.